pyomo/core/expr/logical_expr.py

```diff
@@ -512,4 +531,56 @@ def _apply_operation(self, result):
         return sum(result[1:]) >= result[0]
 
 
+class AllDifferentExpression(NaryBooleanExpression):
+    """
+    Logical expression that all of the N child statements have different values.
+    All arguments are expected to be discrete-valued.
+    """
+
+    __slots__ = ()
+
+    PRECEDENCE = 9  # TODO: maybe?
+
+    def getname(self, *arg, **kwd):
+        return 'all_different'
+
+    def _to_string(self, values, verbose, smap):
+        return "all_different(%s)" % (", ".join(values))
+
+    def _apply_operation(self, result):
+        for val1, val2 in combinations(result, 2):
+            if val1 == val2:
+                return False
+        return True
+
+
+class CountIfExpression(NumericExpression):
+    """
+    Logical expression that returns the number of True child statements.
+    All arguments are expected to be Boolean-valued.
+    """
+
+    __slots__ = ()
+    PRECEDENCE = 10  # TODO: maybe?
+
+    def __init__(self, args):
+        # require a list, a la SumExpression
+        if args.__class__ is not list:
+            args = list(args)
+        self._args_ = args
```

Comment:
Why is this needed for CountIf and not for AllDiff?

Reply:
Whoops, it's not. I don't need to implement my own `__init__` here--just removed it.